zkevm-circuits/src/evm_circuit/witness.rs

```diff
@@ -126,6 +126,7 @@ impl BlockContext {
                         ),
                     ]
                 })
+                .rev()
```

Comment:
This `rev()` function is added.

Reply:
@jeong0982 how does adding this fix the issue of block lookup?

Reply:
I checked it and it still has problem with lookup.. I will fix it

Reply:
Issue of block lookup was not raised because all the values in `history_hashes` are same. I changed values and block lookup error was raised.

Reply:
Do you get the lookup error for HistoryHash? Or for BlockNumber?

Reply:
For `history_hashes` which is defined in test code. 
I don't know why result is different between different `history_hashes`. I thought that values in `history_hashes` are result of lookup, so It couldn't change behavior of lookup. But it isn't.

Reply:
You're getting a lookup error because:
1. Witness for history hash is populated with (block.number - idx - 1). Which means, a reducing index from `current block number` to `current block number - 256`.
2. You're doing a lookup using `block number`

You should do the lookup using `current block number - block number` and it'll work!

Reply:
Is lookup index same as `block_number`? Imo, block hash is stored by reducing index, so `block_number` is in that range(`current_block_number - 256` to `current_block_number`).

Reply:
@jeong0982 the issue is that:
1. geth utils assumes history hashes is in the order `curr_block - 256 -> cur_block - 1`
2. witness generation assumes history hashes is in the order `cur_block - 1 -> cur_block - 256`

The solution was to reverse the history hashes before getting their assignment to the block table. I can push a commit to fix it now.

Reply:
@CPerezz cc ^

Please have a look at the above comment. That's the fix that was needed. In the test case creation, I have used all different history hashes so that we know for sure that the history hash does not clash (giving us an illusion that the lookup is working fine)

Reply:
Thank you for help @roynalnaruto!

Reply:
There's still a test failure, which I'm looking into now

Reply:
Fixed and pushed a commit. The diff between `block_number` and `current_block_number` will be `1 <= diff <= 256`. Hence we have:
```
current_block_number - block_number < 257
```

You were using `256` earlier, hence the issue.

Reply:
Yes, that should be. Thank you :)
I think this conversation can be marked as resolved, right?

Reply:
Just waiting for @CPerezz to review the conversation and commits :)

Reply:
So IIUC, @roynalnaruto fixed the issue changing the witness generation logic right? 
I think this is the correct way, anything that forces us to use `rev()` in opcode impls when we could just change the witness generator is bad IMO.

So as long as I understood the fix correctly. This is fine from my side.